src/common/log.c

```diff
@@ -0,0 +1,318 @@
+ * accepts raw, pre-formatted input and does not use ZF_LOG.
+ */
+static ArdopLogFile SessionLog = {
+	ArdopLogDir,
+	"ARDOPSession",
+	".log",
+	0,
+	0,
+	NULL
+};
+
+static ArdopLogFile* ALL_LOG_FILES[] = {
+	&DebugLog,
+	&SessionLog
+};
+
+static void log_callback_discard(const zf_log_message*, void*) {}
+
+static void log_callback(const zf_log_message* msg, void* param) {
+	(void)param; /* unused */
+
+	const char EOL[] = "\n";
+	const size_t EOL_SZ = sizeof(EOL) - 1;
+
+	// zf_log always leaves enough space in the buffer for a newline
+	memcpy(msg->p, EOL, EOL_SZ);
+
+	if (msg->lvl >= ArdopLogVerbosityConsole)
+	{
+		// write undecorated message to stdout
+#if defined(_WIN32) || defined(_WIN64)
+		/* WriteFile() is atomic for local files opened with FILE_APPEND_DATA and
+		without FILE_WRITE_DATA */
+		DWORD written;
+		WriteFile(GetStdHandle(STD_OUTPUT_HANDLE), msg->msg_b,
+			(DWORD)(msg->p - msg->msg_b + EOL_SZ), &written, 0);
+#else
+		/* write() is atomic for buffers less than or equal to PIPE_BUF. */
+		RETVAL_UNUSED(write(STDOUT_FILENO, msg->msg_b,
+			(size_t)(msg->p - msg->msg_b) + EOL_SZ));
+#endif
+	}
+
+	if (!ArdopLogFileEnabled || msg->lvl < ArdopLogVerbosityFile)
+		return;
+
+	// write decorated message to debug log file
+	ardop_logfile_write(&DebugLog, msg->buf, (size_t)(msg->p - msg->buf) + EOL_SZ);
+
+	// tags for alternate log files
+	switch (msg->tag[0]) {
+		case 'A':
+			// start ARQ session log record
+			ardop_logfile_write(&SessionLog, msg->msg_b, (size_t)(msg->p - msg->msg_b) + EOL_SZ);
+			break;
+		case 'a':
+			// continue ARQ session log record
+			ardop_logfile_write_norotate(&SessionLog, msg->msg_b, (size_t)(msg->p - msg->msg_b) + EOL_SZ);
+			break;
+	}
+}
+
+void ardop_log_start(const bool enable_files) {
+	zf_log_set_output_v(ZF_LOG_PUT_STD, NULL, log_callback);
+	ardop_log_enable_files(enable_files);
+}
+
+void ardop_log_stop() {
+	ardop_log_enable_files(false);
+	zf_log_set_output_v(ZF_LOG_PUT_STD, NULL, log_callback_discard);
+}
+
+void ardop_log_enable_files(const bool file_output) {
+	// Close all log files
+	// They are reopened on the next loggable message
+	for (size_t i = 0; i < sizeof(ALL_LOG_FILES) / sizeof(ALL_LOG_FILES[0]); ++i) {
+		ardop_logfile_close(ALL_LOG_FILES[i]);
+	}
+
+	ArdopLogFileEnabled = file_output;
+}
+
+bool ardop_log_is_enabled_files() {
+	return ArdopLogFileEnabled;
+}
+
+bool ardop_log_set_directory(const char* const logdir) {
+	const char* d = logdir ? logdir : "";
+	int rv = snprintf(ArdopLogDir, sizeof(ArdopLogDir), "%s", d);
+	if (rv < 0 || rv >= sizeof(ArdopLogDir)) {
+		ArdopLogDir[0] = '\0';
+		return false;
+	}
+
+	return true;
+}
+
+const char* ardop_log_get_directory() {
+	return ArdopLogDir;
+}
+
+void ardop_log_set_port(const uint16_t port) {
+	for (size_t i = 0; i < sizeof(ALL_LOG_FILES)/sizeof(ALL_LOG_FILES[0]); ++i) {
+		ALL_LOG_FILES[i]->port = port;
+	}
+}
+
+void ardop_log_set_level_console(const int level) {
```

Comment:
See note in HostInterface.c.  It might be useful to return a value from ardop_log_set_level_console() and ardop_log_set_level_file() to facilitate sending an appropriate response to the host program.

Reply:
40ac8f591748652a6741afd2a459ce8e09fd78f6 migrates from legacy log levels to zf_log levels. A return value proved unnecessary, but we can still add one if you want.

Reply:
I agree.  With the new implementation, it isn't needed.